Reimplement of gate matrix caching

yasuo-ozu opened the pull request:
- Reimplementing gate matrix caching (related to #6)
- Fix  evaluate_state_to_vector bug
- Apply isort and black

- [x] Run tests on m-GPU environment

KKeita27 committed:
add cache to gate conversion for performance improvement (#6)

* add caching

* add cache to simulator.py

* isort and format

Co-authored-by: Keita Kanno <keita.kanno@gmail.com>

yasuo-ozu committed:
Fix linter

yasuo-ozu committed:
Fix evaluate_state_to_vector bug

yasuo-ozu committed:
Fix problem

yasuo-ozu commented:
I finished testing on m-GPU environment

templepmet commented on quri_parts/cuquantum/custatevec/simulator.py:
I'll fix this.  ref: d5d05f793a789601ddd209fd964e83dc52946e5c

templepmet committed:
Give precision to evaluate_state_to_vector

templepmet commented:
I have enabled the specification of precision for `evaluate_to_state_vector`. As a precaution, I will verify the operation on a GPU and confirm that the bug has been resolved (starting from April 14th).

templepmet commented on poetry.lock:
```diff
@@ -0,0 +1,4614 @@
+# This file is automatically @generated by Poetry 2.1.1 and should not be changed by hand.
```
Is adding this file intentional? The repo pins deps via `requirements.txt` and nothing consumes `poetry.lock`, so it looks like an incidental `poetry` byproduct.

templepmet reviewed (changes requested):
@yasuo-ozu I have reviewed. Could you please clarify one point regarding poetry.lock?

templepmet commented:
## GPU verification results (ABCI-Q / H100)

### Cause of the garbage output (e.g. `3.01326628e+23+1.80177665j`) → fixed by this PR
The current `evaluate_state_to_vector` on `main` has a **dtype mismatch**:
- the gate matrix is built with `cp.array(gate_array(g))` (default **complex128**)
- but `apply_matrix` is told it is `CUDA_C_32F` (**complex64**)

cuStateVec reads the memory as the declared type (no implicit cast), so the complex128 buffer is misread as complex64. The garbage value is exactly that byte misread of `1/√2`:

```
1/√2 (complex128) = 0.7071067811865475+0j
raw bytes         = cc3b7f66 9ea0e63f 00000000 00000000   ← 16 bytes
first 8 bytes read as complex64 → 3.0132663e+23 + 1.8017766j   ← matches the observed value
```

This is not environment-specific — the same value reproduces on H100. This PR unifies the dtypes of the state vector, gate matrix, and cuStateVec data/compute types via `precision`, which resolves it.

### Others
- precision fix: matches qulacs on H100 (fidelity ≈ 1.0 for both complex128/64; 16 single-GPU tests pass) ✅
- multi-GPU: sampler/estimator agree with single-GPU on 1/2/4 GPUs ✅ (8 GPUs / n_global_qubits=3 not checked — node topology)

yasuo-ozu committed:
Remove accidentally committed poetry.lock

main does not track poetry.lock; the lock was added by this PR by
mistake. Drop it to keep the branch consistent with the repository.

templepmet reviewed (approved):
LGTM！
